File: library/src/main/java/com/volley/library/flowtag/adapter/BaseFlowAdapter.java
package com.volley.library.flowtag.adapter;

import android.content.Context;
import android.support.annotation.IntRange;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Modifier;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
/**
 * @Describe
 * @Date : 2019-11-29
 * @Email : volleychzm@gmail.com
 * @Author : volley
 */
public abstract class BaseFlowAdapter<T, K extends BaseTagHolder> extends RecyclerView.Adapter<K> {
    protected Context mContext;
    protected int mLayoutResId;
    protected LayoutInflater mLayoutInflater;
    protected List<T> mData;

    public BaseFlowAdapter(@LayoutRes int layoutResId, @Nullable List<T> data) {
        this.mData = data == null ? new ArrayList<T>() : data;
        if (layoutResId != 0) {
            this.mLayoutResId = layoutResId;
        }
    }

    public BaseFlowAdapter(@Nullable List<T> data) {
        this(0, data);
    }

    public BaseFlowAdapter(@LayoutRes int layoutResId) {
        this(layoutResId, null);
    }

    private List<K> mViewHolders = new ArrayList<>();

    public View getBaseItemView(int position, LayoutInflater inflater, ViewGroup parent) {
        if (position < mViewHolders.size()) {
            return mViewHolders.get(position).getConvertView();
        } else {
            View view = inflater.inflate(mLayoutResId, parent, false);
            K holder = createBaseViewHolder(view);
            bindViewHolder(holder, position);
            return view;
        }
    }

    public void setNewData(@Nullable List<T> data) {
        this.mData = data == null ? new ArrayList<T>() : data;
        notifyDataSetChanged();
    }

    @Deprecated
    public void add(@IntRange(from = 0) int position, @NonNull T item) {
        addData(position, item);
    }

    public void addData(@IntRange(from = 0) int position, @NonNull T data) {
        mData.add(position, data);
        notifyItemInserted(position);
        compatibilityDataSizeChanged(1);
    }

    public void addData(@NonNull T data) {
        mData.add(data);
        notifyItemInserted(mData.size());
        compatibilityDataSizeChanged(1);
    }

    public void remove(@IntRange(from = 0) int position) {
        mData.remove(position);
        int internalPosition = position;
        notifyItemRemoved(internalPosition);
        compatibilityDataSizeChanged(0);
        notifyItemRangeChanged(internalPosition, mData.size() - internalPosition);
    }

    public void setData(@IntRange(from = 0) int index, @NonNull T data) {
        mData.set(index, data);
        notifyItemChanged(index);
    }

    public void addData(@IntRange(from = 0) int position, @NonNull Collection<? extends T> newData) {
        mData.addAll(position, newData);
        notifyItemRangeInserted(position, newData.size());
        compatibilityDataSizeChanged(newData.size());
    }

    public void addData(@NonNull Collection<? extends T> newData) {
        mData.addAll(newData);
        notifyItemRangeInserted(mData.size() - newData.size(), newData.size());
        compatibilityDataSizeChanged(newData.size());
    }

    public void replaceData(@NonNull Collection<? extends T> data) {
        if (data != mData) {
            mData.clear();
            mData.addAll(data);
        }
        notifyDataSetChanged();
    }

    private void compatibilityDataSizeChanged(int size) {
        final int dataSize = mData == null ? 0 : mData.size();
        if (dataSize == size) {
            notifyDataSetChanged();
        }
    }

    @NonNull
    public List<T> getData() {
        return mData;
    }

    @Nullable
    public T getItem(@IntRange(from = 0) int position) {
        if (position >= 0 && position < mData.size())
            return mData.get(position);
        else
            return null;
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    @Override
    public int getItemViewType(int position) {
        return getDefItemViewType(position);

    }

    protected int getDefItemViewType(int position) {
        return super.getItemViewType(position);
    }

    @Override
    public K onCreateViewHolder(ViewGroup parent, int viewType) {
        this.mContext = parent.getContext();
        this.mLayoutInflater = LayoutInflater.from(mContext);
        K baseViewHolder = onCreateDefViewHolder(parent, viewType);
        bindViewClickListener(baseViewHolder);
        baseViewHolder.setAdapter(this);
        return baseViewHolder;
    }

    @Override
    public void onViewAttachedToWindow(K holder) {
        super.onViewAttachedToWindow(holder);
    }

    @Override
    public void onAttachedToRecyclerView(final RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
    }

    @Override
    public void onBindViewHolder(K holder, int position) {
        convert(holder, getItem(position));
        mViewHolders.add(holder);
    }

    private void bindViewClickListener(final BaseTagHolder baseViewHolder) {
        if (baseViewHolder == null) {
            return;
        }
        final View view = baseViewHolder.itemView;
        if (view == null) {
            return;
        }
    }


    protected K onCreateDefViewHolder(ViewGroup parent, int viewType) {
        int layoutId = mLayoutResId;
        return createBaseViewHolder(parent, layoutId);
    }

    protected K createBaseViewHolder(ViewGroup parent, int layoutResId) {
        return createBaseViewHolder(getItemView(layoutResId, parent));
    }

    protected K createBaseViewHolder(View view) {
        Class temp = getClass();
        Class z = null;
        while (z == null && null != temp) {
            z = getInstancedGenericKClass(temp);
            temp = temp.getSuperclass();
        }
        K k;
        if (z == null) {
            k = (K) new BaseTagHolder(view);
        } else {
            k = createGenericKInstance(z, view);
        }
        return k != null ? k : (K) new BaseTagHolder(view);
    }

    private K createGenericKInstance(Class z, View view) {
        try {
            Constructor constructor;
            if (z.isMemberClass() && !Modifier.isStatic(z.getModifiers())) {
                constructor = z.getDeclaredConstructor(getClass(), View.class);
                constructor.setAccessible(true);
                return (K) constructor.newInstance(this, view);
            } else {
                constructor = z.getDeclaredConstructor(View.class);
                constructor.setAccessible(true);
                return (K) constructor.newInstance(view);
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }
        return null;
    }

    private Class getInstancedGenericKClass(Class z) {
        Type type = z.getGenericSuperclass();
        if (type instanceof ParameterizedType) {
            Type[] types = ((ParameterizedType) type).getActualTypeArguments();
            for (Type temp : types) {
                if (temp instanceof Class) {
                    Class tempClass = (Class) temp;
                    if (BaseTagHolder.class.isAssignableFrom(tempClass)) {
                        return tempClass;
                    }
                } else if (temp instanceof ParameterizedType) {
                    Type rawType = ((ParameterizedType) temp).getRawType();
                    if (rawType instanceof Class && BaseTagHolder.class.isAssignableFrom((Class<?>) rawType)) {
                        return (Class<?>) rawType;
                    }
                }
            }
        }
        return null;
    }

    protected View getItemView(@LayoutRes int layoutResId, ViewGroup parent) {
        return mLayoutInflater.inflate(layoutResId, parent, false);
    }

    protected abstract void convert(K tagHelper, T item);

    @Override
    public long getItemId(int position) {
        return position;
    }

    private int getItemPosition(T item) {
        return item != null && mData != null && !mData.isEmpty() ? mData.indexOf(item) : -1;
    }

}

File: librarykotlin/src/main/java/com/volley/library/flowtag/adapter/BaseFlowAdapter.kt
package com.volley.library.flowtag.adapter

import android.content.Context
import android.support.annotation.IntRange
import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.volley.library.flowtag.OptionCheck
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Modifier
import java.lang.reflect.ParameterizedType
import java.util.*
/**
 * @Describe
 * @Date : 2019-11-29
 * @Email : volleychzm@gmail.com
 * @Author : volley
 */
abstract class BaseFlowAdapter<T : OptionCheck, K : BaseTagHolder?> @JvmOverloads constructor(@LayoutRes layoutResId: Int, data: List<T>? = null) : RecyclerView.Adapter<K>() {
    protected var mContext: Context? = null
    protected var mLayoutResId = 0
    protected var mLayoutInflater: LayoutInflater? = null
    protected var mData: MutableList<T>

    constructor(data: List<T>?) : this(0, data) {}

    private val mViewHolders: MutableList<K> = ArrayList()
    fun getBaseItemView(position: Int, inflater: LayoutInflater, parent: ViewGroup?): View {
        return if (position < mViewHolders.size) {
            mViewHolders[position]?.getConvertView()!!
        } else {
            val view = inflater.inflate(mLayoutResId, parent, false)
            val holder = createBaseViewHolder(view)
            bindViewHolder(holder, position)
            view
        }
    }

    fun setNewData(data: List<T>?) {
        mData = data as MutableList<T>? ?: ArrayList<T>()
        notifyDataSetChanged()
    }

    @Deprecated("")
    fun add(@IntRange(from = 0) position: Int, item: T) {
        addData(position, item)
    }

    fun addData(@IntRange(from = 0) position: Int, data: T) {
        mData.add(position, data)
        notifyItemInserted(position)
        compatibilityDataSizeChanged(1)
    }

    fun addData(data: T) {
        mData.add(data)
        notifyItemInserted(mData.size)
        compatibilityDataSizeChanged(1)
    }

    fun remove(@IntRange(from = 0) position: Int) {
        mData.removeAt(position)
        notifyItemRemoved(position)
        compatibilityDataSizeChanged(0)
        notifyItemRangeChanged(position, mData.size - position)
    }

    fun setData(@IntRange(from = 0) index: Int, data: T) {
        mData[index] = data
        notifyItemChanged(index)
    }

    fun addData(@IntRange(from = 0) position: Int, newData: Collection<T>) {
        mData.addAll(position, newData)
        notifyItemRangeInserted(position, newData.size)
        compatibilityDataSizeChanged(newData.size)
    }

    fun addData(newData: Collection<T>) {
        mData.addAll(newData)
        notifyItemRangeInserted(mData.size - newData.size, newData.size)
        compatibilityDataSizeChanged(newData.size)
    }

    fun replaceData(data: Collection<T>) {
        if (data !== mData) {
            mData.clear()
            mData.addAll(data)
        }
        notifyDataSetChanged()
    }

    private fun compatibilityDataSizeChanged(size: Int) {
        val dataSize = if (mData == null) 0 else mData.size
        if (dataSize == size) {
            notifyDataSetChanged()
        }
    }

    fun getItem(@IntRange(from = 0) position: Int): T? {
        return if (position >= 0 && position < mData.size) mData[position] else null
    }

    override fun getItemCount(): Int {
        return mData.size
    }

    override fun getItemViewType(position: Int): Int {
        return getDefItemViewType(position)
    }

    protected fun getDefItemViewType(position: Int): Int {
        return super.getItemViewType(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): K {
        mContext = parent.context
        mLayoutInflater = LayoutInflater.from(mContext)
        val baseViewHolder = onCreateDefViewHolder(parent, viewType)
        bindViewClickListener(baseViewHolder)
        baseViewHolder?.setAdapter(this)
        return baseViewHolder
    }

    override fun onViewAttachedToWindow(holder: K) {
        super.onViewAttachedToWindow(holder)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
    }

    override fun onBindViewHolder(holder: K, position: Int) {
        convert(holder, getItem(position))
        mViewHolders.add(holder)
    }

    private fun bindViewClickListener(baseViewHolder: BaseTagHolder?) {
        if (baseViewHolder == null) {
            return
        }
        val view: View = baseViewHolder.itemView ?: return
    }

    protected fun onCreateDefViewHolder(parent: ViewGroup?, viewType: Int): K {
        val layoutId = mLayoutResId
        return createBaseViewHolder(parent, layoutId)
    }

    protected fun createBaseViewHolder(parent: ViewGroup?, layoutResId: Int): K {
        return createBaseViewHolder(getItemView(layoutResId, parent))
    }

    protected fun createBaseViewHolder(view: View): K {
        var temp: Class<*>? = javaClass
        var z: Class<*>? = null
        while (z == null && null != temp) {
            z = getInstancedGenericKClass(temp)
            temp = temp.superclass
        }
        val k: K?
        k = z?.let { createGenericKInstance(it, view) }
                ?: BaseTagHolder(view) as K
        return k ?: BaseTagHolder(view) as K
    }

    private fun createGenericKInstance(z: Class<*>, view: View): K? {
        try {
            val constructor: Constructor<*>
            return if (z.isMemberClass && !Modifier.isStatic(z.modifiers)) {
                constructor = z.getDeclaredConstructor(javaClass, View::class.java)
                constructor.isAccessible = true
                constructor.newInstance(this, view) as K
            } else {
                constructor = z.getDeclaredConstructor(View::class.java)
                constructor.isAccessible = true
                constructor.newInstance(view) as K
            }
        } catch (e: NoSuchMethodException) {
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        } catch (e: InstantiationException) {
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            e.printStackTrace()
        }
        return null
    }

    private fun getInstancedGenericKClass(z: Class<*>): Class<*>? {
        val type = z.genericSuperclass
        if (type is ParameterizedType) {
            val types = type.actualTypeArguments
            for (temp in types) {
                if (temp is Class<*>) {
                    val tempClass = temp
                    if (BaseTagHolder::class.java.isAssignableFrom(tempClass)) {
                        return tempClass
                    }
                } else if (temp is ParameterizedType) {
                    val rawType = temp.rawType
                    if (rawType is Class<*> && BaseTagHolder::class.java.isAssignableFrom(rawType)) {
                        return rawType
                    }
                }
            }
        }
        return null
    }

    protected fun getItemView(@LayoutRes layoutResId: Int, parent: ViewGroup?): View {
        return mLayoutInflater!!.inflate(layoutResId, parent, false)
    }

    protected abstract fun convert(tagHelper: K, item: T?)
    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    private fun getItemPosition(item: T?): Int {
        return if (item != null && mData != null && !mData.isEmpty()) mData.indexOf(item) else -1
    }


    open fun getData(): List<T> {
        return mData
    }


    init {
        mData = data as MutableList<T>? ?: ArrayList<T>()
        if (layoutResId != 0) {
            mLayoutResId = layoutResId
        }
    }
}